petastorm/unischema.py

```diff
@@ -152,24 +152,28 @@ def create_schema_view(self, fields):
         """
 
         # Split fields parameter to regex pattern strings and UnischemaField objects
-        regex_patterns = [f for f in fields if isinstance(f, string_types)]
+        regex_patterns = [field for field in fields if isinstance(field, string_types)]
         # We can not check type against UnischemaField because the artifact introduced by
         # pickling, since depickled UnischemaField are of type collections.UnischemaField
         # while withing depickling they are of petastorm.unischema.UnischemaField
         # Since UnischemaField is a tuple, we check against it since it is invariant to
         # pickling
-        unischema_field_objects = [f for f in fields if isinstance(f, tuple)]
+        unischema_field_objects = [field for field in fields if isinstance(field, tuple)]
         if len(unischema_field_objects) + len(regex_patterns) != len(fields):
             raise ValueError('Elements of "fields" must be either a string (regular expressions) or '
                              'an instance of UnischemaField class.')
 
-        view_fields = unischema_field_objects + match_unischema_fields(self, regex_patterns)
-
-        for field in unischema_field_objects:
-            # Comparing by field names. Prevoiusly was looking for `field not in self._fields.values()`, but it breaks
-            # due to faulty pickling: T223683
-            if field.name not in self._fields:
-                raise ValueError('field {} does not belong to the schema {}'.format(field, self))
+        # For fields that are specified as instances of Unischema: make sure that this schema contains fields
+        # with these names.
+        exact_field_names = [field.name for field in unischema_field_objects]
+        unknown_field_names = set(exact_field_names) - set(self.fields.keys())
```

Comment:
Why are we checking against self.fields.keys() here, shouldn't we be checking against the unischema

Reply:
We use only field names from what user gave to us and not UnischemaField instances to make 'new' codec field does not leak into the actual schema instances used for decoding. Otherwise, we could end up using new codec on data encoded with an old codec.

Reply:
okay, I see